k-distribution/src/main/scripts/lib/pyk/pyk/kcfg.py

```diff
@@ -65,407 +106,335 @@ def __exit__(self, exc_type, exc_value, traceback):
-                result = {}
-                for (key, val) in value.items():
-                    result[key] = KCFG._decode(val)
-        else:
-            result = value
-        return result
-
-    def _assign(self, other):
-        self.states = other.states
-        self.graph = other.graph
-        self.abstractions = other.abstractions
-        self.loop = other.loop
-        self.split = other.split
-        self.init = other.init
-        self.target = other.target
-        self.terminal = other.terminal
-        self.stuck = other.stuck
-        self.subsumptions = other.subsumptions
-        self.frontier = other.frontier
-
-    def to_dict(self):
-        dct = {}
-        for field in KCFG._FIELDS:
-            dct[field] = getattr(self, field)
-        return KCFG._encode(dct)
+    @property
+    def nodes(self) -> Set[Node]:
+        return set(self._nodes.values())
+
+    @property
+    def covers(self) -> Set[Cover]:
+        return set(self._covers.values())
+
+    @property
+    def frontier(self) -> Set[Node]:
+        return {
+            node for node in self._nodes.values()
+            if node.id not in self._edges and node.id not in self._covers and node.id not in self._stuck
+        }
+
+    def to_dict(self) -> Dict[str, Any]:
+        nodes = [node.to_dict() for node in self.nodes]
+        edges = [edge.to_dict() for edge in self.edges()]
+        covers = [cover.to_dict() for cover in self.covers]
+
+        init = list(self._init)
+        target = list(self._target)
+        stuck = list(self._stuck)
+
+        res = {
+            'nodes': nodes,
+            'edges': edges,
+            'covers': covers,
+            'init': init,
+            'target': target,
+            'stuck': stuck,
+        }
+        return {k: v for k, v in res.items() if v}
 
     @staticmethod
-    def from_dict(dct):
-        dct = KCFG._decode(dct)
+    def from_dict(dct: Mapping[str, Any]) -> 'KCFG':
         cfg = KCFG()
-        for field in KCFG._FIELDS:
-            if field in dct:
-                setattr(cfg, field, dct[field])
+
+        nodes: Dict[str, str] = {}
+
+        def resolve(node_id: str) -> str:
+            if node_id not in nodes:
+                raise ValueError(f'Undeclared node: {node_id}')
+            return nodes[node_id]
+
+        for node_dict in dct.get('nodes') or []:
+            term = KInner.from_dict(node_dict['term'])
+            node = cfg.create_node(term)
+
+            node_key = node_dict['id']
+            if node_key in nodes:
+                raise ValueError(f'Multiple declarations of node: {node_key}')
+            nodes[node_key] = node.id
+
+        for edge_dict in dct.get('edges') or []:
+            source_id = resolve(edge_dict['source'])
+            target_id = resolve(edge_dict['target'])
+            condition = KInner.from_dict(edge_dict['condition'])
+            depth = edge_dict['depth']
+            cfg.create_edge(source_id, target_id, condition, depth)
```

Comment:
It seems that we should be able to do something that reads like this instead (not sure it's possible):

```suggestion
        for edge_dict in dct.get('edges') or []:
            cfg.create_edge(Edge.from_dict(edge_dict))
```

The arguments don't quite line up, but this seems cleaner.

Reply:
We could do `cfg._create_edge_from_dict(node_mapping, edge_dict)` where `resolve` (currently a nested function) is a method of `node_mapping` (function `resolve` is necessary in the current implementation as node ids in the dict are not required to be valid hashes.)